c/tskit/trees.c

```diff
@@ -2124,6 +2124,530 @@ tsk_treeseq_sample_count_stat(const tsk_treeseq_t *self, tsk_size_t num_sample_s
+    const tsk_id_t *sample_sets, tsk_size_t num_sample_sets,
+    tsk_bit_array_t *sample_sets_bits)
+{
+    int ret;
+    tsk_bit_array_t bits_row;
+    tsk_size_t j, k, l;
+    tsk_id_t u, sample_index;
+
+    ret = tsk_bit_array_init(sample_sets_bits, self->num_samples, num_sample_sets);
+    if (ret != 0) {
+        return ret;
+    }
+
+    j = 0;
+    for (k = 0; k < num_sample_sets; k++) {
+        tsk_bit_array_get_row(sample_sets_bits, k, &bits_row);
+        for (l = 0; l < sample_set_sizes[k]; l++) {
+            u = sample_sets[j];
+            sample_index = self->sample_index_map[u];
+            if (tsk_bit_array_contains(
+                    &bits_row, (tsk_bit_array_value_t) sample_index)) {
+                ret = TSK_ERR_DUPLICATE_SAMPLE;
+                goto out;
+            }
+            tsk_bit_array_add_bit(&bits_row, (tsk_bit_array_value_t) sample_index);
+            j++;
+        }
+    }
+
+out:
+    return ret;
+}
+
+static int
+tsk_treeseq_two_locus_count_stat(const tsk_treeseq_t *self, tsk_size_t num_sample_sets,
+    const tsk_size_t *sample_set_sizes, const tsk_id_t *sample_sets,
+    tsk_size_t result_dim, const tsk_id_t *set_indexes, general_stat_func_t *f,
+    norm_func_t *norm_f, tsk_size_t num_left_windows, const double *left_windows,
+    tsk_size_t num_right_windows, const double *right_windows, tsk_flags_t options,
+    tsk_size_t *result_size, double **result)
+{
+    // TODO: generalize this function if we ever decide to do weighted two_locus stats.
+    //       We only implement count stats and therefore we don't handle weights.
+    int ret = 0;
+    tsk_bit_array_t sample_sets_bits;
+    bool stat_site = !!(options & TSK_STAT_SITE);
+    bool stat_branch = !!(options & TSK_STAT_BRANCH);
+    double default_windows[] = { 0, self->tables->sequence_length };
+    tsk_size_t state_dim = num_sample_sets;
+    sample_count_stat_params_t f_params = { .sample_sets = sample_sets,
+        .num_sample_sets = num_sample_sets,
+        .sample_set_sizes = sample_set_sizes,
+        .set_indexes = set_indexes };
+
+    tsk_memset(&sample_sets_bits, 0, sizeof(sample_sets_bits));
+
+    ret = tsk_treeseq_check_sample_sets(
+        self, num_sample_sets, sample_set_sizes, sample_sets);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = sample_sets_to_bit_array(
+        self, sample_set_sizes, sample_sets, num_sample_sets, &sample_sets_bits);
+    if (ret != 0) {
+        goto out;
+    }
+
+    /* If no mode is specified, we default to site mode */
+    if (!(stat_site || stat_branch)) {
+        stat_site = true;
+    }
+    /* It's an error to specify more than one mode */
+    if (stat_site + stat_branch > 1) {
+        ret = TSK_ERR_MULTIPLE_STAT_MODES;
+        goto out;
+    }
+
+    if (state_dim < 1) {
+        ret = TSK_ERR_BAD_STATE_DIMS;
+        goto out;
+    }
+    if (result_dim < 1) {
+        ret = TSK_ERR_BAD_RESULT_DIMS;
+        goto out;
+    }
```

Comment:
Since we're not sure about how the ``windows`` thing is going to work I vote we just ``assert(left_windows == NULL && right_windows == NULL)``; for now, and delete the uncovered code.